app/main/routes.py

```diff
@@ -27,16 +29,44 @@ def upload():
             return render_template("upload.html", response=response)
 
         # TODO: Update this to round_four when available
-        response = post_ingest(excel_file, {"source_type": "tf_round_three"})
-        response = response.json()
+        # response = post_ingest(excel_file, {"source_type": "tf_round_three"})
+        # response_a = response.json()
+
+        # TODO: REMOVE THIS ONCE BACK END IS RE-ATTACHED !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+        with open('tests/resources/validation_errors.json') as error_json:
+            response = json.load(error_json)
+        # TODO !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+        # happy path: spreadsheet uploaded successfully
+        if response.get('status_code') == 200:
+            return render_template("success.html", file_name=excel_file.filename)
+
+        # unhappy path pre-validation failure
+        pre_error = response.get('validation_errors').get('PreTransformationErrors')
+        if pre_error:
+            return render_template("upload.html", pre_error=pre_error)
+
+        #unhappy path validation failures
+        tab_errors = response.get('validation_errors').get('TabErrors')
+        if tab_errors:
+
+            # Nested list comprehension
+            # error_items = [{"heading": {"text": tab},
+            #                "content":
+            #                    {"html": f"<h3 class=\"govuk-heading-m\">{error_name}</h3> "
+            #                             f"<p class=\"govuk-body\">{error_desc}</p>"
+            #                     for error_name, error_desc in errors.items()}}
+            #               for tab, errors in tab_errors.items()]
 
-        # TODO: Implement a design system error handler
-        if type(response) is dict:
-            response = response.get("validation_errors")
-        else:
-            response = [response]
+            # for loop with dist comprehension
+            error_items = []
+            for tab, errors in tab_errors.items():
+                error_items.append({"heading": {"text": tab},
+                                    "content": {"html": f"<h3 class=\"govuk-heading-m\">{error_name}</h3> "
+                                                        f"<p class=\"govuk-body\">{error_desc}</p>"
+                                                for error_name, error_desc in errors.items()}})
```

Comment:
Would it be possible to put this logic into a function somewhere else? routes.py should be mostly concerned with routing logic.